community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/Clause.scala

```diff
@@ -96,20 +96,19 @@ case class Match(optional: Boolean, pattern: Pattern, hints: Seq[UsingHint], whe
     def pickFirst(a: Identifier, b: Identifier) = if (a.position.offset < b.position.offset) a else b
 
     @tailrec
-    def loop(compare: Set[Identifier], rest: Seq[Set[Identifier]]) {
-      val shouldWarn = rest.exists( o => (compare intersect o).isEmpty)
-
-      if (shouldWarn) {
-        val notification = CartesianProductNotification(compare.reduce(pickFirst).position)
-        state.notificationLogger += notification
+    def loop(compare: Set[Identifier], rest: Seq[Set[Identifier]], intermediateState: SemanticState): SemanticState = {
+      val updatedState = rest.filter { o =>
+        (compare intersect o).isEmpty
+      }.foldLeft(intermediateState) { (innerState, identifiers) =>
+        innerState.addNotification(CartesianProductNotification(position, identifiers))
       }
 
-      if (rest.nonEmpty) loop(rest.head, rest.tail)
+      if (rest.nonEmpty) loop(rest.head, rest.tail, updatedState) else updatedState
     }
 
-    if (nodes.nonEmpty) loop(nodes.head, nodes.tail)
+    val finalState = if (nodes.nonEmpty) loop(nodes.head, nodes.tail, state) else state
```

Comment:
The difference is that the `SemanticState` previously contained a mutable logger field, but now contains an immutable `Set[InternalNotification]`. Due to that, we can't just add a notification to the state, but we need to keep the return value from the `innerState.addNotification` call and send the state with all notifications back on the last line.